### ISSUE TITLE
KaTeX expressions with escapes and braces throw errors

### ISSUE THREAD

**Issue by hugmanrique** [mdx-js/mdx#370]
### Subject of the issue

I'm trying to setup a Next.js site with mdx and these remark plugins:

- `remark-math@1.0.5`
- `@kwangkim/remark-jsx-katex@1.11.4` (I couldn't find the module's source on GitHub, but you can view it on this [RunKit](https://npm.runkit.com/@kwangkim/remark-jsx-katex))

And this file under `pages/`:

```markdown
This is some extra text that gets rendered correctly if I remove the KaTeX block below:

$$
23_{10} = 2^4+2^2+2^1+2^0 = \underbrace{10111_2}_{\mathrm{binary\ representation}}
$$
```

I assume the escapes inside braces (`{}`) are confusing MDX and it's trying to parse them as a expression, which causes the following error:

![React error](https://i.imgur.com/MIrBSSH.png)

([gist error](https://gist.github.com/hugmanrique/dc3fb9a7978d84362e13037f8d3a3763))

As you can see, it's complaining about a "bad character escape sequence". One way to fix this would be to ignore all the braces and escapes `{}\` inside `$` (inlineMath) or `$$` (math) blocks.

### Your environment

- Windows 10 Pro 1803
- `@zeit/next-mdx`: 1.2.0
  - `@mdx-js/mdx`: 0.15.7
- `next`: 7.0.2-canary.49

### Steps to reproduce

Create new Next.js project with the following `next.config.js` file:

```
const remarkMath = require('remark-math');
const katex = require('@kwangkim/remark-jsx-katex');

module.exports = require('@zeit/next-mdx')({
  options: {
    mdPlugins: [remarkMath, katex],
  },
  pageExtensions: ['js', 'mdx']
});
```

Now, create `pages/foo.mdx`:

```markdown
This is some extra text that gets rendered correctly if I remove the KaTeX block below:

$$
23_{10} = 2^4+2^2+2^1+2^0 = \underbrace{10111_2}_{\mathrm{binary\ representation}}
$$
```

Now, run `next` and load `localhost:3000/foo` in your browser.

### Expected behaviour

The KaTeX escapes should be ignored even if they are inside braces (`{}`).

### Actual behaviour

The KaTeX escapes and generally, all the content inside braces doesn't get ignored, causing compiling and React rendering issues.

**Comment by ChristopherBiscardi** [comment]
We ran into this in gatsby-mdx too (actually we sort of run into things like this a lot because of the way the gatsby plugins function). I added some [replacement code](https://github.com/ChristopherBiscardi/gatsby-mdx/blob/affcfe7855f60c97fe726fb4ee1ae7eebb791719/packages/gatsby-mdx/utils/babel-plugin-html-attr-to-jsx-attr.js#L72-L87) in a babel plugin that we use to post-process MDX content after remark and hast plugins have run but before we ship the JSX.

```js
if (
  path.node.openingElement.name.name ===
  "annotation" /* && if gatsby-remark-katex is enabled */
) {
  const genCode = path.node.children.map(ast => generate.default(ast).code);
  path.node.children = [
    t.jSXText(
      genCode
        .join("")
        .replace("{", "&#123;")
        .replace("}", "&#125;")
    )
  ];
}
```

> As you can see, it's complaining about a "bad character escape sequence". One way to fix this would be to ignore all the braces and escapes {}\ inside $ (inlineMath) or $$ (math) blocks.

There's a similar "javascript string escaping"-related issue in https://github.com/mdx-js/mdx/pull/366. so I wonder if we should keep adding to the list of "non-escaped sections" (code blocks, `$` or `$$` sequences, etc) or somehow default to "html style" interpretation for special chars like `{`, `\u`, etc and force `\{` or something for bracket literals to get passed through into the final jsx? dunno, haven't had a moment to think through the ramifications here all the way.

**Comment by ChristopherBiscardi** [comment]
> And it didn't work.

yeah, probably because in gatsby-mdx we [apply it to the MDX output before returning it for the user's babel config to run on](https://github.com/ChristopherBiscardi/gatsby-mdx/blob/affcfe7855f60c97fe726fb4ee1ae7eebb791719/packages/gatsby-mdx/loaders/mdx-loader.js#L170). It's used as some internal infrastructure, not as a plugin inline with whatever's in `next/babel` (for example). The equivalent place for next looks like it would be [between the mdx loader and the babel loader](https://github.com/zeit/next-plugins/blob/8d58b4147e37f081ff76058dc8be9f4e29c1c37c/packages/next-mdx/index.js#L15-L19) in next/babel, but it also looks like that's not configurable to the user.

If I had to guess, I'd assume that the JSX expressions are already compiled to function calls by something in the preset before the plugin even gets to run, thus no JSX expressions means no logs.

> On another note, I think I found a bug in your babel plugin implementation:

cool, thanks for the heads up. iirc, I only manually tested it on a file that had one expression in it and there aren't any tests on the plugin yet because I still haven't released the 0.3.x line while I try to discover more processing issues.

**Comment by hugmanrique** [comment]
I finally found out the cause were the backward slashes (`\`) and mdx seems to handle the braces fine (sorry for the false statement above, this was already fixed in #145). I've created a short rehype plugin that fixes the issue by double escaping this character:

```js
const visit = require('unist-util-visit');
const remarkMath = require('remark-math');
const katex = require('rehype-katex');

module.exports = require('@zeit/next-mdx')({
  options: {
    mdPlugins: [remarkMath],
    hastPlugins: [
      katex,
      () => tree =>
        visit(tree, 'element', node => {
          if (node.tagName !== 'annotation') {
            return;
          }

          if (!node.properties || node.properties.encoding !== 'application/x-tex') {
            return;
          }

          const { value } = node.children[0];

          // Double escape backward slashes
          node.children[0].value = value.replace(/\\/g, '\\\\');
        })
    ]
  },
  pageExtensions: ['js', 'mdx']
})
```

The escapes get removed by KaTeX when rendering, but it leaves the original TeX parsed string in an `<annotation encoding="application/x-tex">` element.

**Comment by johno** [comment]
This should now be fixed in `0.20.2`.
